File: .github/actions/toc-action/insert_toc.py
#!/usr/bin/env python3
import argparse, re, sys
from pathlib import Path
from glob import glob
from collections import defaultdict
import unicodedata

MARKER_START = "<!-- START doctoc generated TOC please keep comment here to allow auto update -->"
MARKER_END = "<!-- END doctoc generated TOC please keep comment here to allow auto update -->"

def github_slug(text, seen):
    # Normalize and remove combining marks
    s = unicodedata.normalize("NFKD", text)
    s = "".join(ch for ch in s if not unicodedata.combining(ch))
    s = s.strip().lower()
    # remove punctuation except spaces and unicode letters/numbers
    s = re.sub(r"[^\w\s-]", "", s, flags=re.UNICODE)
    s = re.sub(r"\s+", "-", s)
    s = s.strip("-")
    base = s or "-"
    count = seen[base]
    seen[base] += 1
    return base if count == 0 else f"{base}-{count}"

def build_toc(headings, collapsed):
    if not headings:
        return ""
    seen = defaultdict(int)
    lines = []
    for level, text in headings:
        indent = "  " * max(0, level - 2)
        anchor = github_slug(text, seen)
        lines.append(f"{indent}- [{text}](#{anchor})")
    toc = "\n".join(lines)
    if collapsed:
        return f"<details>\n<summary>Table of Contents</summary>\n\n{toc}\n\n</details>"
    return toc

def process_file(path, collapsed):
    txt = path.read_text(encoding="utf-8")
    if MARKER_START not in txt or MARKER_END not in txt:
        print(f"Skipping {path}: markers not found")
        return False
    # collect headings >= level 2
    headings = []
    for m in re.finditer(r'^(#{2,6})\s+(.*)$', txt, flags=re.MULTILINE):
        lvl = len(m.group(1))
        txt_h = m.group(2).strip()
        headings.append((lvl, txt_h))
    toc = build_toc(headings, collapsed)
    before, rest = txt.split(MARKER_START, 1)
    _, after = rest.split(MARKER_END, 1)
    new = before + MARKER_START + "\n\n" + toc + "\n\n" + MARKER_END + after
    if new != txt:
        path.write_text(new, encoding="utf-8")
        print(f"Updated {path}")
        return True
    print(f"No changes for {path}")
    return False

def expand_files(patterns):
    out = []
    for p in patterns:
        p = p.strip()
        if any(ch in p for ch in "*?[]"):
            out += glob(p, recursive=True)
        else:
            out.append(p)
    return [Path(x) for x in out]

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--files", default="./README.md")
    ap.add_argument("--collapsed", default="true")
    args = ap.parse_args()
    patterns = args.files.split(",")
    collapsed = args.collapsed.lower() in ("1","true","yes")
    paths = expand_files(patterns)
    changed = False
    for p in paths:
        if not p.exists():
            print(f"Not found: {p}")
            continue
        if process_file(p, collapsed):
            changed = True
    sys.exit(0 if changed else 0)

if __name__ == "__main__":
    main()

File: .github/actions/toc-action/insert-toc.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");

const argv = require("minimist")(process.argv.slice(2), {
  string: ["files", "collapsed"],
  default: { files: "./README.md", collapsed: "true" },
});

const filesArg = argv.files;
const collapsed = String(argv.collapsed).toLowerCase() === "true";
const markers = {
  start:
    "<!-- START doctoc generated TOC please keep comment here to allow auto update -->",
  end: "<!-- END doctoc generated TOC please keep comment here to allow auto update -->",
};

function slugify(text) {
  return text
    .trim()
    .toLowerCase()
    .replace(/<\/?[^>]+(>|$)/g, "")
    .replace(/[^\w\s-]/g, "")
    .replace(/\s+/g, "-");
}

function buildToc(headings) {
  if (headings.length === 0) return "";
  const lines = headings.map((h) => {
    const indent = "  ".repeat(Math.max(0, h.level - 2));
    const anchor = slugify(h.text);
    return `${indent}- [${h.text}](#${anchor})`;
  });
  const toc = lines.join("\n");
  if (collapsed) {
    return `\n---\n<details>\n<summary>Table of Contents</summary>\n\n${toc}\n\n</details>\n---\n`;
  } else {
    return toc;
  }
}

function processFile(filePath) {
  const content = fs.readFileSync(filePath, "utf8");
  const startIdx = content.indexOf(markers.start);
  const endIdx = content.indexOf(markers.end);

  if (startIdx === -1 || endIdx === -1 || endIdx < startIdx) {
    console.log(`Skipping ${filePath}: TOC markers not found or malformed`);
    return false;
  }

  // Extract headings level >= 2
  const lines = content.split(/\r?\n/);
  const headings = [];
  for (const line of lines) {
    const m = line.match(/^(#{2,6})\s+(.*)$/);
    if (m) {
      const level = m[1].length;
      const text = m[2].trim();
      headings.push({ level, text });
    }
  }

  const toc = buildToc(headings);
  const before = content.slice(0, startIdx + markers.start.length);
  const after = content.slice(endIdx);
  const newContent = `${before}\n\n${toc}\n\n${after}`;

  if (newContent !== content) {
    fs.writeFileSync(filePath, newContent, "utf8");
    console.log(`Updated TOC in ${filePath}`);
    return true;
  } else {
    console.log(`No changes for ${filePath}`);
    return false;
  }
}

function expandFiles(arg) {
  // simple comma split and trim; support single file or comma list
  return arg
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
}

const files = expandFiles(filesArg);
let changed = false;
for (const f of files) {
  const p = path.resolve(process.cwd(), f);
  if (!fs.existsSync(p)) {
    console.warn(`File not found: ${p}`);
    continue;
  }
  const didChange = processFile(p);
  changed = changed || didChange;
}

if (changed) {
  process.exitCode = 0;
} else {
  process.exitCode = 0;
}
